### PR TITLE
AttentionNet: A Simple Attention-based Model for Healthcare Predictive Tasks

### PR DESCRIPTION
**Pull request by nikhilsarin1**
# AttentionNet: A Simple Attention-based Model for Healthcare Predictive Tasks

## Author Information
- **Name**: Nikhil Sarin
- **NetID**: nsarin2

## Contribution Type
Model Implementation

## Description
This pull request introduces a new model called AttentionNet, a simple attention-based neural network for healthcare predictive tasks. The model uses a self-attention mechanism to focus on important aspects of patient health data for various prediction tasks like readmission prediction, mortality prediction, etc.

AttentionNet is designed as a general-purpose model that can be used with any of the existing PyHealth tasks including readmission prediction, mortality prediction, drug recommendation, and length of stay prediction. It follows the standard PyHealth model structure and API, making it easy to integrate with existing pipelines.

The key features of AttentionNet include:
- Multi-head self-attention mechanism
- Support for all PyHealth task modes (binary, multiclass, multilabel, regression)
- Customizable embedding and hidden dimensions
- Built-in dropout for regularization
- Mean and sum pooling operations

## Implementation Details
The implementation includes two new files:
1. `pyhealth/models/attention_net.py` - Contains the AttentionNet model implementation
2. `pyhealth/unittests/test_attention_net.py` - Contains unit tests for the model

The model implementation follows the PyHealth coding standards and documentation requirements. Unlike the previous models that inherit from BaseModel, this implementation follows the current structure of PyHealth's models, directly inheriting from nn.Module and implementing the necessary methods to work with the PyHealth framework.

## How to Test
You can test the model using the provided unit tests:

```bash
python -m unittest pyhealth.unittests.test_attention_net
```

The tests verify:
1. Proper initialization of the model
2. Successful forward pass with synthetic data
3. Correct prediction function behavior with appropriate activation functions

For practical use, you can use the model with any of the existing PyHealth datasets and tasks, following standard PyHealth patterns:

```python
from pyhealth.datasets import MIMIC3Dataset
from pyhealth.tasks import readmission_prediction_mimic3_fn
from pyhealth.models import AttentionNet
from pyhealth.datasets import split_by_patient, get_dataloader
from pyhealth.trainer import Trainer

# Load dataset and set task
mimic3base = MIMIC3Dataset(...)
mimic3sample = mimic3base.set_task(task_fn=readmission_prediction_mimic3_fn)
train_ds, val_ds, test_ds = split_by_patient(mimic3sample, [0.8, 0.1, 0.1])

# Create dataloaders
train_loader = get_dataloader(train_ds, batch_size=32, shuffle=True)
val_loader = get_dataloader(val_ds, batch_size=32, shuffle=False)
test_loader = get_dataloader(test_ds, batch_size=32, shuffle=False)

# Initialize model
model = AttentionNet(
    dataset=mimic3sample,
    feature_keys=["conditions", "procedures"],
    label_keys=["label"],
    embedding_dim=128,
    hidden_dim=256,
    attention_heads=4,
    dropout=0.5,
)

# Train and evaluate
trainer = Trainer(model=model)
trainer.train(
    train_dataloader=train_loader,
    val_dataloader=val_loader,
    epochs=50,
)
metrics = trainer.evaluate(test_loader)
```

## Relevance
This implementation is highly relevant to healthcare AI as attention mechanisms have been shown to be effective in many healthcare applications. They can help models focus on the most important features in a patient's medical history, leading to better predictive performance. The attention mechanism is particularly useful for modeling longitudinal patient data where the importance of different medical events may vary across time.

Attention mechanisms have been widely adopted in healthcare ML research, including for:
1. Electronic health record (EHR) analysis
2. Medical time series prediction
3. Medical imaging analysis
4. Clinical note interpretation

The model is designed to be general-purpose and can be used for any of the existing tasks in PyHealth, making it a valuable addition to the library's model collection.

### CODE REVIEW

**Comment by jhnwu3** [comment]
Closing this PR as it lacks proper labeling. Please add appropriate labels and reopen if needed.
